aiida_vasp/calcs/workchains/converge.py

```diff
@@ -0,0 +1,1369 @@
+# pylint: disable=too-many-lines, too-many-locals, too-many-statements, attribute-defined-outside-init
+"""
+ConvergenceWorkChain.
+
+Intended to be used to control convergence checks for plane-wave calculations.
+"""
+import copy
+import numpy as np
+
+from aiida.work.workchain import WorkChain, append_, while_, if_
+from aiida.common.extendeddicts import AttributeDict
+from aiida.orm import WorkflowFactory, Code
+
+from aiida_vasp.utils.aiida_utils import get_data_class, \
+    displaced_structure, compressed_structure, new_structure, new_parameter, new_kpoints
+from aiida_vasp.calcs.workchains.auxiliary.utils import fetch_k_grid, init_input
+
+
+class ConvergeWorkChain(WorkChain):
+    """A workchain to perform convergence tests."""
+
+    _verbose = True
+    _next_workchain = WorkflowFactory('vasp.relax')
+
+    # a few default values defined here for now
+    _encut_start_default = 200
+    _encut_step_default = 50
+    _encut_samples_default = 3
+    _k_step_default = 0.15
+    _k_spacing_default = 0.5
+    _k_samples_default = 3
+    _cutoff_type_default = 'energy'
+    _cutoff_value_default = 0.001
+    _cutoff_value_r_default = 0.001
+    _compress_default = False
+    _displacement_default = False
+    _displacement_vector_default = np.array([1.0, 1.0, 1.0])
+    _displacement_distance_default = 0.2  # AA
+    _displacement_atom_default = 1  # atom number starting from 1
+    _volume_change_default = np.array([[1.05, 0.0, 0.0], [0.0, 1.05, 0.0], [0.0, 0.0, 1.05]])
+    _relax_default = False
+
+    _ALLOWED_CUTOFF_TYPES = {'energy': 0, 'forces': 1, 'vbm': 2, 'gap': 3}
+
+    @classmethod
+    def define(cls, spec):
+        super(ConvergeWorkChain, cls).define(spec)
+        spec.input('code', valid_type=Code)
+        spec.input('structure', valid_type=(get_data_class('structure'), get_data_class('cif')))
+        spec.input('potential_family', valid_type=get_data_class('str'))
+        spec.input('potential_mapping', valid_type=get_data_class('parameter'))
+        spec.input('incar', valid_type=get_data_class('parameter'))
+        spec.input('options', valid_type=get_data_class('parameter'))
+        spec.input('kpoints', valid_type=get_data_class('array.kpoints'), required=False)
+        spec.input('settings', valid_type=get_data_class('parameter'), required=False)
+        spec.input('restart.max_iterations', valid_type=get_data_class('int'), required=False)
+        spec.input('restart.clean_workdir', valid_type=get_data_class('bool'), required=False)
+        spec.input('verify.max_iterations', valid_type=get_data_class('int'), required=False)
+        spec.input('verify.clean_workdir', valid_type=get_data_class('bool'), required=False)
+        spec.input('relax.incar', valid_type=get_data_class('parameter'), required=False)
+        spec.input('relax.perform', valid_type=get_data_class('bool'), required=False, default=get_data_class('bool')(True))
```

Comment:
Consider using `get_data_node('bool', True)` instead of `get_data_class('bool')(True)`, unless you find the latter more clear / readable.

Same goes for all calls to `get_data_class(..)(..)`

Reply:
I have no strong preference here. I think both are readable. How about Aiida and other plugin stuff, is that class()(), or class('','')? Guess we should follow what is already established. You decide.

Reply:
Aiida only has the `DataFactory(class_name)(*args, **kwargs)`, however inside `aiida-vasp` we have so far been consistently using `get_data_node(class_name, *args, **kwargs)` instead (mainly for convenience in typing), at least since it's introduction. And there is a chance it might be adopted in AiiDA in the future.

For consistency within the plugin, I would suggest to adopt `get_data_node`, given it is a simple search/replace operation.

Reply:
Agreed. I will go through everything and make sure this is used. Should be just remove `get_data_class`?

Reply:
No, because sometimes we really need the class, for `valid_type=get_data_class(...)` for instance, or for `isinstance` comparisons.

Reply:
Yes, I forgot about that.